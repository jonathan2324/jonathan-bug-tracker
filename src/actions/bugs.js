import uuid from 'uuid'
import database from '../firebase/firebase'

//ADD_BUG
export const addBug = (bug) => ({
    type: 'ADD_BUG',
    bug
    
})

//START addBUg
export const startAddBug = (bugData = {}) => {
    return (dispatch) => {
        const { 
            name = '',
            description = '',
            notes = '',
            priority = 'Low',
            completed = false,
            createdAt = 0,
            contributions = []
          } = bugData

          const bug = { name, description, priority, completed, notes, createdAt, contributions }

          return database.ref('bugs').push(bug).then((ref) => {
              dispatch(addBug({
                  id: ref.key,
                  ...bug
              }))
          })
    }
}

//REMOVE_BUG

export const removeBug = ({ id }) => ({
    type: 'REMOVE_BUG',
    id
})

//Edit Bug
export const editBug = (id, updates) => ({
    type: 'EDIT_BUG',
    id,
    updates
})

//TOGGLE COMPLETED
export const toggleCompletedBug = (id, completed) => ({
    type: 'TOGGLE_COMPLETED_BUG',
    id,
    completed
})

//ADD_CONTRIBUTION
export const addContribution = (id, {
    name = '',
    contribution = ''
}) => ({
    type: 'ADD_CONTRIBUTION',
    id,
    contribution:
    {
        id: uuid(),
        name,
        contribution
    }

})

//EDIT_CONTRIBUTION
export const editContribution = (bugID, contributionID, updatedContribution) => ({
    type: 'EDIT_CONTRIBUTION',
    bugID,
    contributionID,
    updatedContribution
})

//REMOVE_CONTRIBUTION
export const removeContribution = (bugID, contributionID) => ({
    type: 'REMOVE_CONTRIBUTION',
    bugID,
    contributionID
})