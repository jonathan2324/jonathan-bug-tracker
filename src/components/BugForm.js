import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'



const now = moment()
console.log(now.format('MMM Do YYYY'))

class BugForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            name: props.bug ? props.bug.name : '',
            priority: props.bug ? props.bug.priority : 'Low',
            description: props.bug ? props.bug.description : '',
            notes: props.bug ? props.bug.notes : '',
            createdAt: props.bug ? moment(props.bug.createdAt) : moment(),
            calendarFocused: false,
            error: ''
    
        }
    }
    

    onNameChange = (e) => {
        const name = e.target.value 
        this.setState(() => ({ name }))
    }

    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }

    onNotesChange = (e) => {
        const notes = e.target.value
        this.setState(() => ({ notes }))       
    }

    onPriorityChange = (e) => {
        const priority = e.target.value
        this.setState(() => ({ priority }))
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({
                createdAt
            }))

        }

    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }

    onSubmit = (e) => {
        e.preventDefault()

        if (!this.state.name || !this.state.description) {
            this.setState(() => ({ error: 'Please provide an employee name and a description'}))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                description: this.state.description,
                notes: this.state.notes,
                name: this.state.name,
                createdAt: this.state.createdAt.valueOf(),
                priority: this.state.priority
            })
        }
    }

    render () {
        return (
            <div>
                <p>{this.state.error}</p>
                <form onSubmit={this.onSubmit}>
                    <input type='text' placeholder='Employee name' autoFocus value={this.state.name} onChange={this.onNameChange}/>
                    <select value={this.state.priority} onChange={this.onPriorityChange}>
                        <option value="High">High</option>
                        <option value="Med">Med</option>
                        <option value="Low">Low</option>
                    </select>

                    <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    />

                    <textarea type='text' maxLength="50" placeholder='Enter error description (max length 50 characters)' value={this.state.description} onChange={this.onDescriptionChange}/>
                    <textarea type='text' placeholder='Enter important notes about error' value={this.state.notes} onChange={this.onNotesChange}/>
                    <button>Add Bug</button>

                </form>
            </div>
        )
    }
}

export default BugForm