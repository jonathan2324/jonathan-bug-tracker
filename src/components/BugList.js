import React from 'react'
import { connect } from 'react-redux'
import BugListItem from './BugListItem'
import selectBugs from '../selectors/bugs'

export const BugList = (props) => (
    <div>
        {props.uncompletedBugs.length > 0 && <h3>Uncompleted Bugs [{props.uncompletedBugs.length}]:</h3>}
        {
            props.bugs.length === 0 ? (

                <p>No bugs</p>

            ) : (
            
                props.uncompletedBugs.map((bug) => {
                    return <BugListItem key={bug.id} {...bug}/>
                 })

            )
        }
        {props.completedBugs.length > 0 && <h3>Completed bugs [{props.completedBugs.length}]:</h3>}
        {
            props.bugs.length === 0 ? (

                <p></p>

            ) : (
            
                props.completedBugs.map((bug) => {
                    return <BugListItem key={bug.id} {...bug}/>
                 })

            )
        }



    </div>
)

const mapStateToProps = (state) => {
    const visibleBugs = selectBugs(state.bugs, state.filters)

    return {
        bugs: visibleBugs,
        completedBugs: visibleBugs.filter((bug) => {return bug.completed === true}),
        uncompletedBugs: visibleBugs.filter((bug) => {return bug.completed === false})

    }
}

const ConnectedBugList = connect(mapStateToProps)(BugList)

export default ConnectedBugList