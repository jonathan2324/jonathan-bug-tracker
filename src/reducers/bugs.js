

//bugReducer
const bugReducerDefaultState = []

const bugReducer = (state = bugReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_BUG':
            return [
                ...state,
                action.bug
            ]
        case 'REMOVE_BUG':
            return state.filter(({ id }) => id !== action.id)
        case 'EDIT_BUG':
            return state.map((bug) => {
                if (bug.id === action.id) {
                    return {
                        ...bug,
                        ...action.updates
                    }
                } else {
                    return bug
                }

            })
        case 'TOGGLE_COMPLETED_BUG':
            return state.map((bug) => {
                if (bug.id === action.id) {
                    return {
                        ...bug,
                        completed: !action.completed
                    }
                } else {
                    return bug
                }
            })
        case 'ADD_CONTRIBUTION':
            return state.map((bug) => {
                if(bug.id === action.id){
                    return {
                        ...bug,
                        contributions: [
                            
                            action.contribution,
                            ...bug.contributions,
                            
                        ]
                    }
                } else return bug
            })
        default:
            return state;
    }
};

export default bugReducer