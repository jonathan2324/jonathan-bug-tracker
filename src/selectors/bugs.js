import moment from 'moment'

//Get visible bugs
const getVisibleBugs = (bugs, { text, sortBy, startDate, endDate }) => {
    return bugs.filter((bug) => {
        const createdAtMoment = moment(bug.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'): true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        const textMatch = bug.name.toLowerCase().includes(text.toLowerCase())
        
        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if(sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        }
        else {
            return null
        }
    }).sort((a,b) => {
       if(sortBy === 'completed') {
           return a.completed ? 1 : -1
       } else {
           return 0
       }
    }).sort((a, b) => {
        if(sortBy === 'priority') {
            if(a.priority === 'High') {
                return b.priority === 'Med' || b.priority === 'Low' ? -1 : 1
            }
            else if (a.priority === 'Med') {
                return b.priority === 'High' ? 1 : -1
            } else {
                return 0
            }
        }
            
    })
}

export default getVisibleBugs