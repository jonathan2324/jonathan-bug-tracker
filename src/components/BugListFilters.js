import React from 'react'
import {connect} from 'react-redux'
import { setTextFilter, sortByDate, sortByPriority, setStartDate, setEndDate, sortByCompleted, filterDefault, filterByCompleted, filterByUncompleted } from '../actions/filters'
import { DateRangePicker } from 'react-dates'
import getVisibleBugs from '../selectors/bugs'



export class BugListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({startDate, endDate}) => { //destructure object that is passed into this function by the date range picker
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)        //these are the startDate and endDate that was required by airbnb
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused: calendarFocused }))
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
        }
    
    onSortChange = (e) => {
        if (e.target.value === "date") {
            this.props.sortByDate()
        } else if (e.target.value === 'priority') {
            this.props.sortByPriority()
        } 
    }

    onFilterChange = (e) => {
        if(e.target.value === 'uncompleted') {
            this.props.filterByUncompleted()
        } else if (e.target.value === 'completed') {
            this.props.filterByCompleted()
        } else if (e.target.value === 'all') {
            this.props.filterDefault()
        }
    }

    render() {
        return (
            <div>
                <input type='text' value={this.props.filters.text} onChange={this.onTextChange}/>
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value='date'>Date</option>
                    <option value='priority'>Priority</option>
                    
                </select>
                <select value={this.props.filters.filterBy} onChange={this.onFilterChange}>
                    <option value='all'>View all</option>
                    <option value='uncompleted'>View open</option>
                    <option value='completed'>View closed</option>
                </select>
                <DateRangePicker 
                startDate={this.props.filters.startDate} 
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => false}
                />
            </div>
        )
    }

}

const mapStateToProps = (state) => {
    const visibleBugs = getVisibleBugs(state.bugs, state.filters)
    return {
        filters: state.filters,
        completedBugs: visibleBugs.filter((bug) => { return bug.completed === 'true'}),
        uncompletedBugs: visibleBugs.filter((bug) => { return bug.completed === 'false'})
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByPriority: () => dispatch(sortByPriority()),
    sortByCompleted: () => dispatch(sortByCompleted()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    filterDefault: () => dispatch(filterDefault()),
    filterByCompleted: () => dispatch(filterByCompleted()),
    filterByUncompleted: () => dispatch(filterByUncompleted())
})

export default connect(mapStateToProps, mapDispatchToProps)(BugListFilters)