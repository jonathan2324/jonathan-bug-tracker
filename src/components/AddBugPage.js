import React from 'react';
import BugForm from './BugForm'
import { connect } from 'react-redux'
import { startAddBug } from '../actions/bugs'

export class AddBugPage extends React.Component {
  onSubmit = (bug) => {
    this.props.startAddBug(bug)
    this.props.history.push('/')
  }
  render() {
    return (
      <div>
        <h1>Add bug</h1>
        <BugForm onSubmit={this.onSubmit}/>
      </div>
    )
  }
}


const mapDispatchToProps = (dispatch) => {
  return {
    startAddBug:(bug) => dispatch(startAddBug(bug))
  }
}

export default connect(undefined, mapDispatchToProps)(AddBugPage);