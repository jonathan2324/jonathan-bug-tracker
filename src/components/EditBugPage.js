import React from 'react';
import { connect } from 'react-redux'
import BugForm from '../components/BugForm'
import { removeBug } from '../actions/bugs'
import { editBug } from '../actions/bugs';

export class EditBugPage extends React.Component {
  //method for onSubmit using editBug  dispatch action 
  onSubmit = (bugFormObject) => { //this bugFormObject variable is from the BugForm where we use this.props.onSubmit({}) and created a bug object
    this.props.editBug(this.props.bug.id, bugFormObject)  //need to use this variable because it represents the updated object but the id comes from mapStateToProps
    this.props.history.push('/')
  }
  //method for onRemove using removeBug dispatch action
  onRemove = (e) => {
    this.props.removeBug({ id: this.props.bug.id })
    this.props.history.push('/')

  }

  render() {

  
  return (
    <div >
      <h2>Ticket number: {this.props.bug.id.substr(0,8)}</h2>
      <BugForm 
      
      bug={this.props.bug} //This represents the bug from mapStateToProps, this is used in BugForm when we check if props.bug ? or use default
      onSubmit={this.onSubmit}/>
      <button 
        
        onClick={this.onRemove}>Remove</button>
    </div>
  );
}
}


const mapStateToProps = (state, props) => {
  return {
    bug: state.bugs.find((bug) => bug.id === props.match.params.id)
  }
}

const mapDispatchToProps = (dispatch, props) => {
  return {
    editBug: (id, bugFormObj) => dispatch(editBug(id, bugFormObj)),
    removeBug: (data) => dispatch(removeBug(data)) //we can call the variable being passed in whatever we want 
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditBugPage);