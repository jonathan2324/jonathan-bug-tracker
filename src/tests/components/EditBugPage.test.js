import React from 'react'
import { shallow } from 'enzyme'
import {EditBugPage} from '../../components/EditBugPage'
import bugs from '../fixtures/bugs'
import toJSON from 'enzyme-to-json'


let editBug, startRemoveBug, history, wrapper
beforeEach(() => {
    editBug = jest.fn()
    startRemoveBug = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(<EditBugPage bug={bugs[0]} editBug={editBug} startRemoveBug={startRemoveBug} history={history}/>)
})

test('Should render EditBugPage', () => {
    expect(toJSON(wrapper)).toMatchSnapshot()

})

test('Should handle editBug', () => {
    wrapper.find('BugForm').prop('onSubmit')(bugs[0])
    expect(editBug).toHaveBeenLastCalledWith( bugs[0].id ,bugs[0])
    expect(history.push).toHaveBeenLastCalledWith('/')
})


test('Should handle startRemoveBug', () => {
    wrapper.find('button').simulate('click')
    expect(startRemoveBug).toHaveBeenLastCalledWith({ id: bugs[0].id })
    expect(history.push).toHaveBeenLastCalledWith('/')
})