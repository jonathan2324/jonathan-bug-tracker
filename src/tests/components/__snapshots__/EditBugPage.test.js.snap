// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditBugPage 1`] = `
<div>
  <h2>
    Ticket number: 
    1
  </h2>
  <BugForm
    bug={
      Object {
        "completed": true,
        "createdAt": 0,
        "description": "Error in X",
        "id": "1",
        "name": "Lebron",
        "notes": "notes 1",
        "priority": "Med",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
  <ContributionsForm
    bug={
      Object {
        "completed": true,
        "createdAt": 0,
        "description": "Error in X",
        "id": "1",
        "name": "Lebron",
        "notes": "notes 1",
        "priority": "Med",
      }
    }
    onSubmitContribution={[Function]}
  />
  <ContributionsList
    bug={
      Object {
        "completed": true,
        "createdAt": 0,
        "description": "Error in X",
        "id": "1",
        "name": "Lebron",
        "notes": "notes 1",
        "priority": "Med",
      }
    }
  />
</div>
`;
